go/vt/vtgate/planbuilder/testdata/reference_cases.json

```diff
@@ -0,0 +1,680 @@
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source"
+      ]
+    },
+    "gen4-plan": {
+      "Instructions": {
+        "Keyspace": {
+          "Name": "main",
+          "Sharded": false
+        },
+        "MultiShardAutocommit": false,
+        "OperatorType": "Delete",
+        "Query": "delete from ambiguous_ref_with_source where col = 1",
+        "Table": "ambiguous_ref_with_source",
+        "TargetTabletType": "PRIMARY",
+        "Variant": "Unsharded"
+      },
+      "Original": "delete from ambiguous_ref_with_source where col = 1",
+      "QueryType": "DELETE",
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source"
+      ]
+    }
+  },
+  {
+    "comment": "delete from qualified ambiguous reference table error no primary vindex",
+    "query": "delete from user.ambiguous_ref_with_source where col = 1",
+    "v3-plan": "table 'ambiguous_ref_with_source' does not have a primary vindex",
+    "gen4-plan": {
+      "Instructions": {
+        "Keyspace": {
+          "Name": "main",
+          "Sharded": false
+        },
+        "MultiShardAutocommit": false,
+        "OperatorType": "Delete",
+        "Query": "delete from ambiguous_ref_with_source where col = 1",
+        "Table": "ambiguous_ref_with_source",
+        "TargetTabletType": "PRIMARY",
+        "Variant": "Scatter"
+      },
+      "Original": "delete from user.ambiguous_ref_with_source where col = 1",
+      "QueryType": "DELETE",
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source"
+      ]
+    }
+  },
+  {
+    "comment": "join with unqualified unambiguous ref with source routes to requested table",
+    "query": "select user.col from user join ref_with_source",
+    "v3-plan": {
+      "Instructions": {
+        "FieldQuery": "select `user`.col from `user` join ref_with_source where 1 != 1",
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "OperatorType": "Route",
+        "Query": "select `user`.col from `user` join ref_with_source",
+        "Table": "`user`, ref_with_source",
+        "Variant": "Scatter"
+      },
+      "Original": "select user.col from user join ref_with_source",
+      "QueryType": "SELECT"
+    },
+    "gen4-plan": {
+      "Instructions": {
+        "FieldQuery": "select `user`.col from `user`, ref_with_source where 1 != 1",
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "OperatorType": "Route",
+        "Query": "select `user`.col from `user`, ref_with_source",
+        "Table": "`user`, ref_with_source",
+        "Variant": "Scatter"
+      },
+      "Original": "select user.col from user join ref_with_source",
+      "QueryType": "SELECT",
+      "TablesUsed": [
+        "user.ref_with_source",
+        "user.user"
+      ]
+    }
+  },
+  {
+    "comment": "join with unqualified reference does not optimize routes if source & reference different names; TODO(maxeng) not working",
```

Comment:
My intention with the PR was to only allow optimal reference routing to happen when both the reference and source have the same table name. It looks like that constraint isn't in place.

TODO: put constraint in place, unless a PR reviewer says they prefer the current behavior.

Reply:
The name should have nothing to do with optimal routing, it should just be determined if any reference table is present in the current keyspace for the source table or not.

I like this behaviour, I would have pushed for this if any constraint would have placed :)

Reply:
I would say you can change the test comment.

Reply:
Done in 8c8da64fe1